### PR TITLE
Add CLI params support + MobileNetV2 model + small refactoring

### PR DESCRIPTION
**Pull request by rpoltorak**
- It allows to run experiments easily using CLI params like for example:
```bash
python mask_classifier.py --gpus 1 --image_size 256 --batch_size 32 --learning_rate 0.001
```
and many many more.

- Moved basic cnn model and data set into separate directories to create space for other experiments in the future
- Fixed accuracy calculation and modified some model layers (just for testing purposes - it's still dummy)
- Added MobileNetV2 network with customized fully connected layer for testing purposes

### CODE REVIEW

File: mask_detector/datasets/masked_face_net.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    def __getitem__(self, key):
+        row = self.dataFrame.iloc[key]
+
+        mask = torch.tensor([row['mask']], dtype=torch.float)
+        image = Image.open(row['image'])
+
+        return self.transform(image), mask
+
+    def __len__(self):
+        return len(self.dataFrame.index)
```
Comment (GrivIN) [inline mask_detector/datasets/masked_face_net.py:33]: you can check len of DataFrame (skipping index)
```suggestion
        return len(self.dataFrame)
```

File: mask_detector/datasets/masked_face_net.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import pandas as pd
+
+from PIL import Image
+from torch.utils.data.dataset import Dataset
+from torchvision.transforms import CenterCrop, Compose, Normalize, Resize, ToPILImage, ToTensor
+
+
+class MaskedFaceNetDataset(Dataset):
+    def __init__(self, csv_file, image_size):
+        self.dataFrame = pd.read_csv(csv_file)
```
Comment (GrivIN) [inline mask_detector/datasets/masked_face_net.py:15]: nit: can we use snake_case style?

File: mask_detector/mask_classifier.py [hunk 3]
```diff
@@ -5,85 +5,95 @@
...
-from model import BasicCNN
-from dataset import MaskDataset
 from utils import train_val_test_split
+from models.basic_cnn import BasicCNN
+from models.mobile_net_v2 import MobileNetV2
+from datasets.masked_face_net import MaskedFaceNetDataset
 
 
 class MaskClassifier(LightningModule):
     def __init__(self, net, learning_rate=1e-3):
```
Comment (GrivIN) [inline mask_detector/mask_classifier.py:19]: can we use something readable instead `1e-3`

File: mask_detector/mask_classifier.py [hunk 4]
```diff
@@ -5,85 +5,95 @@
...
     def training_step(self, batch, batch_idx):
         x, y = batch
         out = self.net(x)
-
         loss = binary_cross_entropy(out, y)
 
-        self.log('train_loss', loss, on_epoch=True)
-
-        return loss
+        return {'loss': loss, 'accuracy': self.accuracy(out, y)}
```
Comment (GrivIN) [inline mask_detector/mask_classifier.py:33]: I do not like dictionaries returned from functions, can we use data classes here or it is a framework requirement to have dicts here?

File: mask_detector/mask_classifier.py [hunk 5]
```diff
@@ -5,85 +5,95 @@
...
 
     def validation_step(self, batch, batch_idx):
         x, y = batch
         out = self.net(x)
         loss = binary_cross_entropy(out, y)
 
-        self.log('valid_loss', loss, on_step=True)
+        return {'loss': loss, 'accuracy': self.accuracy(out, y)}
```
Comment (GrivIN) [inline mask_detector/mask_classifier.py:40]: dataclasses are possible here?

File: mask_detector/mask_classifier.py [hunk 6]
```diff
@@ -5,85 +5,95 @@
...
         x, y = batch
         out = self.net(x)
         loss = binary_cross_entropy(out, y)
 
-        _, out = torch.max(out, dim=1)
-        val_acc = accuracy_score(out.cpu(), y.cpu())
-        val_acc = torch.tensor(val_acc)
-
-        return {'test_loss': loss, 'test_acc': val_acc}
+        return {'loss': loss, 'accuracy': self.accuracy(out, y)}
```
Comment (GrivIN) [inline mask_detector/mask_classifier.py:47]: dataclasses are possible here?

**Review by GrivIN — CHANGES_REQUESTED** [review]
Actually, couple of comments
btw. can we require type annotations?

File: mask_detector/models/mobile_net_v2.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            ReLU(),
+            Linear(in_features=128, out_features=1),
+            Sigmoid()
+        )
+
+    def forward(self, x):
+        with torch.no_grad():
+            x = self.features(x)
+
+        x = self.classifier(x)
```
Comment (jedrzejpolaczek) [inline mask_detector/models/mobile_net_v2.py:28]: Would be nice to have more meaningful names for variables.

File: mask_detector/models/mobile_net_v2.py [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import torch
```
Comment (jedrzejpolaczek) [inline mask_detector/models/mobile_net_v2.py:1]: Nice and clean!

File: mask_detector/datasets/masked_face_net.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
+""" Dataset module
+"""
```
Comment (jedrzejpolaczek) [inline mask_detector/datasets/masked_face_net.py:2]: I also do not get the name. Why `masked face net`?

**Review by jedrzejpolaczek — CHANGES_REQUESTED** [review]

File: mask_detector/models/basic_cnn.py [hunk 10]
```diff
@@ -8,35 +8,35 @@ def __init__(self):
...
 
         self.linearLayers = Sequential(
-            Linear(in_features=128*25*25, out_features=2048),
-            ReLU(),
-            Linear(in_features=2048, out_features=1024),
+            Flatten(),
+            Linear(in_features=128*15*15, out_features=1024),
             ReLU(),
             Linear(in_features=1024, out_features=1),
-            Softmax()
```
[the remark this answers is not in the document]
Comment (rpoltorak) [reply mask_detector/models/basic_cnn.py:32]: Because It's better to use Sigmoid function to binary classification problems. Softmax is better for multi-classification. This is the interesting article that explains a bit: http://web.stanford.edu/~nanbhas//blog/sigmoid-softmax.html

File: mask_detector/mask_classifier.py (hunk 6 above)
Comment (rpoltorak) [reply mask_detector/mask_classifier.py]: Maybe, but this is not necessary at this moment because I removed this and return only loss value (everything else should be logged independently by using `self.log()` method.

File: mask_detector/mask_classifier.py (hunk 5 above)
Comment (rpoltorak) [reply mask_detector/mask_classifier.py]: Maybe, but this is not necessary at this moment because I removed this and return only loss value (everything else should be logged independently by using self.log() method.

File: mask_detector/mask_classifier.py (hunk 4 above)
Comment (rpoltorak) [reply mask_detector/mask_classifier.py]: Maybe, but this is not necessary at this moment because I removed this and return only loss value (everything else should be logged independently by using self.log() method.

File: mask_detector/mask_classifier.py (hunk 3 above)
Comment (rpoltorak) [reply mask_detector/mask_classifier.py]: Done :+1:

File: mask_detector/datasets/masked_face_net.py (hunk 2 above)
Comment (rpoltorak) [reply mask_detector/datasets/masked_face_net.py]: Of course :)

File: mask_detector/datasets/masked_face_net.py (hunk 1 above)
Comment (rpoltorak) [reply mask_detector/datasets/masked_face_net.py]: Good point! Changed. I'm pretty sure that using `len(self.data_frame.index)` is a little bit more performant but `len(self.data_frame)` is definitely much cleaner.

File: mask_detector/datasets/masked_face_net.py [hunk 11]
```diff
@@ -0,0 +1,33 @@
+""" Dataset module
```
[the remark this answers is not in the document]
Comment (rpoltorak) [reply mask_detector/datasets/masked_face_net.py:1]: That's good point, you're right. However, it sounds like a separate task, so I'm gonna create a separate task for that.

File: mask_detector/datasets/masked_face_net.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+from torch.utils.data.dataset import Dataset
+from torchvision.transforms import CenterCrop, Compose, Normalize, Resize, ToPILImage, ToTensor
+
+
+class MaskedFaceNetDataset(Dataset):
+    def __init__(self, csv_file, image_size):
+        self.dataFrame = pd.read_csv(csv_file)
+
+        self.transform = Compose([
+            Resize(image_size),  # for MobileNetV2 - set image size to 256
```
[the remark this answers is not in the document]
Comment (rpoltorak) [reply mask_detector/datasets/masked_face_net.py:18]: This is indeed a good one. Do you have any idea how we could name that? We need to think about naming that kind of combinations of dataset and model. I can't come up with a name for a combination of MobileNetV2 and MaskedFaceNet :)

File: mask_detector/datasets/masked_face_net.py (hunk 9 above)
Comment (rpoltorak) [reply mask_detector/datasets/masked_face_net.py:2]: Because it's the name of our current dataset. MaskedFaceNet :)

File: mask_detector/models/basic_cnn.py [hunk 13]
```diff
@@ -1,4 +1,4 @@
-from torch.nn import Module, Conv2d, Linear, MaxPool2d, ReLU, Sequential, Softmax, Flatten
+from torch.nn import Dropout, Module, Conv2d, Flatten, Linear, MaxPool2d, ReLU, Sequential, Sigmoid, Softmax, Flatten
 
 
 class BasicCNN(Module):
```
[the remark this answers is not in the document]
Comment (rpoltorak) [reply mask_detector/models/basic_cnn.py:4]: I love that idea. Created separate README.md file in models directory. You could check this here: `mask_detector/models/README.md`

File: mask_detector/models/mobile_net_v2.py (hunk 7 above)
Comment (rpoltorak) [reply mask_detector/models/mobile_net_v2.py:28]: That's totally true. Changed to `output`. Hope it's clear now :)

File: mask_detector/mask_classifier.py [hunk 14]
```diff
@@ -5,86 +5,107 @@
...
 
-from model import BasicCNN
-from dataset import MaskDataset
 from utils import train_val_test_split
+from models.basic_cnn import BasicCNN
+from models.mobile_net_v2 import MobileNetV2
+from datasets.masked_face_net import MaskedFaceNetDataset
 
 
 class MaskClassifier(LightningModule):
```
Comment (azjawinski) [inline mask_detector/mask_classifier.py:17]: Hmm, probably it would be convenient to prepare some basic, semi-generic class for our DataModule.
It should define all the common operations for the task we have (image-processing), as probably most of them will be the same between different models.

Then we should extend this base class and overwrite specific methods for each Model we have (choosing correct datasets, defining some transformations etc).

What do you think?

File: mask_detector/models/README.md [hunk 15]
```diff
@@ -0,0 +1,7 @@
+# Models
+
+## Current performance measures
+
+| Model       | Params                                                | Loss   | Recall |
+| ----------- | ----------------------------------------------------- | ------ | ------ |
+| MobileNetV2 | max_epochs=10<br>batch_size=32<br>learning_rate=0.001 | 0.1449 | 0.9794 |
```
Comment (azjawinski) [inline mask_detector/models/README.md:7]: Having only loss and recall can be very misleading without having the precision as well.

[Recall does not inform about any false positives](https://en.wikipedia.org/wiki/Precision_and_recall). I suppose you have checked the dataset and if classes are properly balanced.

We have low loss value - though it does not mean anything. We do not know how to interpret this, or is loss `0.1` much better than `0.2`. This is why usually additionally `accuracy` is presented.
On its own accuracy is very risky metric (without precision and recall data we could have a 99% accuracy when model is e.g. always outputting the same value, but classes were not balanced). 

However, it describes what is the most important in evaluating the model outcome - how accurate the model is in the most intuitive way. If we see Recall / Precision / F1 score of e.g. 0.7 we need to think how to interpret it relating to the dataset, possible predictions etc. If we have an accuracy of 0.8 we already see that 80% of all predictions are correct. **Then** we can investigate other metrics to directly compare models, but that accuracy is the true **goal** anyways.

Also we miss **precision**. Here is a nice presentation what this could lead to (based on the situation from a shooting range):
image

I would opt for tracking all of these metrics (as they are not so slow to compute anyways). We never know when such information could become helpful.

File: mask_detector/models/basic_cnn.py [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+            ReLU(),
+            Linear(in_features=1024, out_features=1),
+            Sigmoid()
+        )
+
+    def forward(self, output):
+        output = self.convLayers1(output)
+        output = self.convLayers2(output)
+        output = self.convLayers3(output)
+        output = self.linearLayers(output)
```
Comment (azjawinski) [inline mask_detector/models/basic_cnn.py:41]: We could try to investigate if this generates some calculation efficiency cost. There could be the case that PyTorch is optimizing all the calculations in 1 Sequential layer, that we lose keeping separate Sequentials. Maybe it would be better to store them as arrays and then convert to 1 Sequential?

TBC (To Be Checked) though, it may be not a problem at all.

**Review by azjawinski — APPROVED** [review]
Looks great 💙 

Added some picky comments, but probably some of them will be fixed when I finally propose some collab tool and the project structure 😞 Because of this I'm already giving approve ✅

**Comment by KonradNojman** [comment]
@rpoltorak Will you be adding something more to this PR, or we can merge it and @azjawinski will work on this topic in the new PRs?

**Comment by azjawinski** [comment]
Closing this PR, I'm taking this over after Remek.
Comments will be implemented (and further reviewed) in a separate PR.
